FormalConjectures/Wikipedia/SnakeInTheBox.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright 2025 The Formal Conjectures Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-/
+
+import FormalConjectures.Util.ProblemImports
+
+/-!
+# Snake in the box
+
+*References:*
+- [Wikipedia](https://en.wikipedia.org/wiki/Snake-in-the-box)
+- [xkcd](https://xkcd.com/3125/)
+-/
+
+universe u
+
+open SimpleGraph
+
+def Hypercube (n : ℕ) : SimpleGraph (Finset (Fin n)) := fromRel fun a b => (a ∩ b).card = 1
+
+def IsSnakeInGraphOfLength {V : Type u} [DecidableEq V] (G : SimpleGraph V) (G' : Subgraph G)
+    (k : ℕ) : Prop :=
+  G'.IsInduced ∧ ∃ u v : V, ∃ (P : G.Walk u v), P.IsPath ∧ G'.verts = P.support.toFinset.toSet ∧
+  P.length = k
+
+noncomputable def LongestSnakeInGraph {V : Type u} [DecidableEq V] (G : SimpleGraph V) : ℕ :=
```

Comment:
I think it would be good to have docstrings for these three definitions!

Reply:
done